File: packages/libs/SubstrateML/Sources/SubstrateML/TokenCompaction.swift
// TokenCompaction.swift
//
// The token-compaction transform (SPEC_DISTILLATION_STORAGE §7.6) and the
// token-count estimator (§6). ONE pure transform, two uses:
//
//   • the short-item distillation path (§7.5) — items under 3 sentences
//     get their distilled rendering by compacting the content directly;
//   • the tokenized-on-read hydration variants (§10.1) —
//     `content_tokenized` / `distilled_tokenized` are computed at
//     retrieval time by this same transform and never stored.
//
// The matrix-path Stage 5 rendering (§7.4) also compacts each unit
// sentence through this transform.
//
// DETERMINISM CONTRACT (§5.3 rule 6): given the same input string the
// output is byte-identical across runs and bit-identical between this
// implementation and the Rust twin (token_compaction.rs). Everything here
// is fixed-table, integer, and ASCII-case arithmetic — no locale, no ICU,
// no Foundation string APIs whose behavior could differ across platforms.
// Conformance vectors: TokenCompactionConformanceTests (Swift) and
// token_compaction_conformance.rs (Rust) assert identical outputs.
//
// FORMAT RULES implemented (SPEC §5.3, priority order):
//   1. Propositional fidelity — negation and quantifiers ALWAYS survive
//      (they are never in the drop tables); entities, numbers, units,
//      dates pass through verbatim (a token containing a digit or an
//      uppercase first letter never matches the lowercase drop tables).
//   2. Stopword/filler removal — a fixed, deliberately CONSERVATIVE list:
//      only words whose removal cannot flip or ambiguate a proposition
//      (articles, intensifier fillers, pleasantries, bare present
//      copulas). Ambiguous words ("well", "just", "mean") are excluded
//      because they have noun/adjective readings rule 1 protects.
//   3. Dense wording — a fixed phrase-rewrite table (longest match wins).
//      Perfect-passive auxiliaries compress ("has been moved" → "moved");
//      tense-CHANGING rewrites (e.g. "will be" → "") are deliberately
//      absent because rule 1 outranks rule 3.
//   4. Minimal markup — decorative Unicode punctuation normalizes to
//      ASCII; no markdown is produced.
//   5. Single spacing — whitespace runs collapse; a dropped token's
//      trailing clause punctuation migrates to the previous token so
//      clause structure survives.

import Foundation

/// The Phase 1 distillation format + pipeline contract identifier
/// (SPEC §4). Stored in `distilled_pipeline_version`; a row whose stored
/// value differs from this constant is a regeneration candidate for the
/// sweep. Bump this string whenever the rendering contract changes
/// (compaction tables, Stage 5 ordering, estimator formula) — stored
/// renderings regenerate lazily against the new contract.
public enum DistillationPipelineVersion {
    /// Phase 1 contract: TokenCompaction tables v1 + Stage 5 core-first
    /// ordering + the (3B + 16W + 12)/24 token estimator, all pinned to
    /// `DistillationPipeline.defaultExtractor` (the extractor present and
    /// bit-identical on both legs).
    public static let current = "p1"
}

/// The §7.6 token-compaction transform and §6 token estimator. No
/// instances — a namespace of pure functions.
public enum TokenCompaction {

    // MARK: - Fixed tables (v1 — changing ANY entry is a pipeline-version bump)

    /// Single-word drops (rule 2), matched against the lowercase word core.
    /// Conservative by construction: articles, unambiguous intensifier
    /// fillers, pleasantries, discourse adverbs, and the bare present
    /// copulas. NO negations, NO quantifiers, NO modality words
    /// (maybe/might/possibly change a proposition's force and stay),
    /// NO past copulas (was/were carry tense).
    static let stopwords: Set<String> = [
        // articles
        "a", "an", "the",
        // intensifier fillers
        "really", "very", "quite", "actually", "basically", "literally",
        "honestly", "frankly",
        // discourse
        "anyway",
        // pleasantries
        "please",
        // bare present copulas (identity/property assertions keep their
        // meaning telegraphically: "color is blue" → "color blue")
        "is", "are", "am",
    ]

    /// Multi-word rewrites (rule 3), matched on lowercase word cores,
    /// longest match first at each position. Replacement words are
    /// emitted lowercase; sentence recapitalization repairs casing.
    /// An empty replacement drops the phrase (perfect-passive
    /// auxiliaries, filler idioms).
    static let phrases: [(match: [String], replacement: [String])] = [
        (["due", "to", "the", "fact", "that"], ["because"]),
        (["at", "this", "point", "in", "time"], ["now"]),
        (["in", "the", "event", "that"], ["if"]),
        (["as", "a", "result", "of"], ["because", "of"]),
        (["with", "regard", "to"], ["regarding"]),
        (["in", "order", "to"], ["to"]),
        (["make", "sure", "to"], []),
        (["has", "been"], []),
        (["have", "been"], []),
        (["had", "been"], []),
        (["kind", "of"], []),
        (["sort", "of"], []),
        (["of", "course"], []),
    ]

    /// ASCII clause punctuation that migrates when its carrier token drops.
    private static let clausePunctuation: Set<Character> = [".", ",", ";", ":", "!", "?"]

    // MARK: - compact

    /// Apply the §7.6 transform. Pure; deterministic; Rust-bit-identical.
    public static func compact(_ text: String) -> String {
        // Rule 4: normalize decorative Unicode punctuation to ASCII first
        // so the tokenizer and tables see one spelling.
        let normalized = normalizeUnicodePunctuation(text)

        // Rule 5: tokenize on any whitespace run (spaces, tabs, newlines,
        // blank-line runs all collapse).
        let tokens = normalized.split(whereSeparator: { $0.isWhitespace }).map(String.init)
        guard !tokens.isEmpty else { return "" }

        // Word cores (lowercased, ASCII case fold only) for table matching.
        let cores = tokens.map { lowercaseCore(of: $0) }

        var out: [String] = []
        var i = 0
        while i < tokens.count {
            // Rule 3: longest phrase match at this position.
            if let (matchLen, replacement) = phraseMatch(cores: cores, at: i) {
                // Trailing clause punctuation of the LAST matched token
                // survives the rewrite (the clause boundary is real content).
                let trailing = trailingPunctuation(of: tokens[i + matchLen - 1])
                if replacement.isEmpty {
                    migrate(punctuation: trailing, onto: &out)
                } else {
                    var words = replacement
                    words[words.count - 1] += trailing
                    out.append(contentsOf: words)
                }
                i += matchLen
                continue
            }
            // Rule 2: single-word drop, guarded by rule 1 (a token whose
            // core contains a digit or an uppercase letter never matches
            // the lowercase tables, so numbers/dates/entities survive).
            if stopwords.contains(cores[i]) {
                migrate(punctuation: trailingPunctuation(of: tokens[i]), onto: &out)
                i += 1
                continue
            }
            out.append(tokens[i])
            i += 1
        }

        // Rule 5: single spaces; then repair sentence-initial casing that
        // article drops exposed ("the plan works." → "plan works." →
        // "Plan works."). ASCII-only uppercase so both legs agree.
        return recapitalizeSentences(out.joined(separator: " "))
    }

    // MARK: - estimateTokenCount (§6)

    /// Approximate BPE token count of `text` on cl100k-class vocabularies
    /// for English-dominant prose. Integer fixed-point blend of the two
    /// classic estimators (bytes/4 and words*4/3), averaged and rounded:
    ///
    ///     est = (3*B + 16*W + 12) / 24     (integer division)
    ///
    /// where B = UTF-8 byte count and W = whitespace-separated word count.
    /// Chosen because dense stopword-stripped prose runs closer to one
    /// token per word than chars/4 alone predicts; the blend lands within
    /// the ±20% §6 accuracy target on the conformance corpus. Advisory
    /// only — never load-bearing. Bit-identical in the Rust twin (pure
    /// integer arithmetic).
    public static func estimateTokenCount(_ text: String) -> Int64 {
        let words = text.split(whereSeparator: { $0.isWhitespace })
        guard !words.isEmpty else { return 0 }
        let bytes = Int64(text.utf8.count)
        let wordCount = Int64(words.count)
        return (3 * bytes + 16 * wordCount + 12) / 24
    }

    // MARK: - Internals

    /// Fixed decorative-Unicode → ASCII map (rule 4). Only unambiguous
    /// typographic variants are mapped; everything else passes through.
    static func normalizeUnicodePunctuation(_ text: String) -> String {
        var out = String()
        out.reserveCapacity(text.count)
        for ch in text {
            switch ch {
            case "\u{2018}", "\u{2019}", "\u{201B}": out.append("'")
            case "\u{201C}", "\u{201D}", "\u{201F}": out.append("\"")
            case "\u{2013}", "\u{2014}", "\u{2015}": out.append("-")
            case "\u{2026}": out.append("...")
            case "\u{00A0}", "\u{2007}", "\u{202F}": out.append(" ")
            default: out.append(ch)
            }
        }
        return out
    }

    /// The token's word core: leading/trailing ASCII punctuation stripped,
    /// ASCII letters lowercased (ASCII-only fold — no locale). Internal
    /// apostrophes/hyphens stay, so "doesn't" cores to "doesn't" and never
    /// matches a drop table (negation survival, rule 1).
    static func lowercaseCore(of token: String) -> String {
        let scalars = Array(token.unicodeScalars)
        var start = 0
        var end = scalars.count
        while start < end, isStrippablePunct(scalars[start]) { start += 1 }
        while end > start, isStrippablePunct(scalars[end - 1]) { end -= 1 }
        var out = String.UnicodeScalarView()
        for s in scalars[start..<end] {
            if s.value >= 65 && s.value <= 90 { // A-Z → a-z
                out.append(Unicode.Scalar(s.value + 32)!)
            } else {
                out.append(s)
            }
        }
        return String(out)
    }

    /// ASCII punctuation strippable at token edges for core extraction.
    private static func isStrippablePunct(_ s: Unicode.Scalar) -> Bool {
        switch s {
        case ".", ",", ";", ":", "!", "?", "\"", "'", "(", ")", "[", "]",
             "{", "}", "-", "*", "_", "`":
            return true
        default:
            return false
        }
    }

    /// The token's trailing run of clause punctuation (".,;:!?").
    private static func trailingPunctuation(of token: String) -> String {
        var suffix = ""
        for ch in token.reversed() {
            if clausePunctuation.contains(ch) {
                suffix.insert(ch, at: suffix.startIndex)
            } else {
                break
            }
        }
        return suffix
    }

    /// Migrate a dropped token's trailing clause punctuation onto the
    /// previous emitted token — unless it already ends with clause
    /// punctuation (no doubling, rule 5) or there is no previous token.
    private static func migrate(punctuation: String, onto out: inout [String]) {
        guard !punctuation.isEmpty, let last = out.last, let lastChar = last.last,
              !clausePunctuation.contains(lastChar) else { return }
        out[out.count - 1] += punctuation
    }

    /// Uppercase the character at each sentence start IF it is an ASCII
    /// lowercase letter — the repair for casing that article drops expose
    /// ("the plan works." → "plan works." → "Plan works."). A sentence
    /// start is the beginning of the string, or the first non-whitespace
    /// character after a sentence terminator. A terminator is ".", "!",
    /// or "?" followed by whitespace/end — but NOT a dot inside an
    /// ellipsis ("wait... done" does not start a new sentence) and NOT a
    /// dot inside a number ("3.5"). A boundary whose first character is
    /// not a lowercase ASCII letter (a quote, a digit, an entity already
    /// capitalized) is left untouched. ASCII-only case arithmetic so both
    /// legs agree byte-for-byte.
    static func recapitalizeSentences(_ text: String) -> String {
        var scalars = Array(text.unicodeScalars)
        var atSentenceStart = true
        for i in 0..<scalars.count {
            let s = scalars[i]
            let isWhitespace = s == " " || s == "\t" || s == "\n" || s == "\r"
            if atSentenceStart && !isWhitespace {
                if s.value >= 97 && s.value <= 122 { // a-z
                    scalars[i] = Unicode.Scalar(s.value - 32)! // → A-Z
                }
                atSentenceStart = false
            }
            // Terminator detection with look-around: ".", "!", "?" ends a
            // sentence only when followed by whitespace or end-of-string,
            // and a "." preceded by another "." is an ellipsis member,
            // not a terminator.
            if s == "." || s == "!" || s == "?" {
                let followedByBreak = i + 1 == scalars.count
                    || scalars[i + 1] == " " || scalars[i + 1] == "\t"
                    || scalars[i + 1] == "\n" || scalars[i + 1] == "\r"
                let ellipsisMember = s == "." && i > 0 && scalars[i - 1] == "."
                if followedByBreak && !ellipsisMember {
                    atSentenceStart = true
                }
            }
        }
        var out = String.UnicodeScalarView()
        for s in scalars { out.append(s) }
        return String(out)
    }

    /// Longest phrase-table match at token position `i`, or nil.
    private static func phraseMatch(
        cores: [String], at i: Int
    ) -> (length: Int, replacement: [String])? {
        var best: (length: Int, replacement: [String])? = nil
        for (match, replacement) in phrases {
            let len = match.count
            guard i + len <= cores.count else { continue }
            guard best == nil || len > best!.length else { continue }
            var matched = true
            for k in 0..<len where cores[i + k] != match[k] {
                matched = false
                break
            }
            if matched {
                best = (len, replacement)
            }
        }
        return best
    }
}

File: packages/libs/SubstrateML/rust/src/token_compaction.rs
// token_compaction.rs
//
// The token-compaction transform (SPEC_DISTILLATION_STORAGE §7.6) and the
// token-count estimator (§6). Rust twin of TokenCompaction.swift —
// BIT-IDENTICAL output required for every input (conformance vectors in
// token_compaction_conformance.rs mirror TokenCompactionConformanceTests).
//
// Everything here is fixed-table, integer, and ASCII-case arithmetic — no
// locale, no Unicode case folding beyond the fixed decorative-punctuation
// map. See the Swift file for the full rule rationale (§5.3 priority
// order; conservative drop tables; fidelity guards).

/// The Phase 1 distillation format + pipeline contract identifier
/// (SPEC §4). Mirrors Swift `DistillationPipelineVersion.current`.
pub const DISTILLATION_PIPELINE_VERSION: &str = "p1";

/// Single-word drops (rule 2), matched against the lowercase word core.
/// MUST stay identical to Swift `TokenCompaction.stopwords`.
const STOPWORDS: &[&str] = &[
    // articles
    "a", "an", "the",
    // intensifier fillers
    "really", "very", "quite", "actually", "basically", "literally",
    "honestly", "frankly",
    // discourse
    "anyway",
    // pleasantries
    "please",
    // bare present copulas
    "is", "are", "am",
];

/// Multi-word rewrites (rule 3), longest match first at each position.
/// MUST stay identical to Swift `TokenCompaction.phrases`.
const PHRASES: &[(&[&str], &[&str])] = &[
    (&["due", "to", "the", "fact", "that"], &["because"]),
    (&["at", "this", "point", "in", "time"], &["now"]),
    (&["in", "the", "event", "that"], &["if"]),
    (&["as", "a", "result", "of"], &["because", "of"]),
    (&["with", "regard", "to"], &["regarding"]),
    (&["in", "order", "to"], &["to"]),
    (&["make", "sure", "to"], &[]),
    (&["has", "been"], &[]),
    (&["have", "been"], &[]),
    (&["had", "been"], &[]),
    (&["kind", "of"], &[]),
    (&["sort", "of"], &[]),
    (&["of", "course"], &[]),
];

fn is_clause_punct(c: char) -> bool {
    matches!(c, '.' | ',' | ';' | ':' | '!' | '?')
}

fn is_strippable_punct(c: char) -> bool {
    matches!(
        c,
        '.' | ',' | ';' | ':' | '!' | '?' | '"' | '\'' | '(' | ')' | '[' | ']'
            | '{' | '}' | '-' | '*' | '_' | '`'
    )
}

/// Apply the §7.6 transform. Pure; deterministic; Swift-bit-identical.
pub fn compact(text: &str) -> String {
    let normalized = normalize_unicode_punctuation(text);
    let tokens: Vec<&str> = normalized.split_whitespace().collect();
    if tokens.is_empty() {
        return String::new();
    }
    let cores: Vec<String> = tokens.iter().map(|t| lowercase_core(t)).collect();

    let mut out: Vec<String> = Vec::new();
    let mut i = 0;
    while i < tokens.len() {
        // Rule 3: longest phrase match at this position.
        if let Some((match_len, replacement)) = phrase_match(&cores, i) {
            let trailing = trailing_punctuation(tokens[i + match_len - 1]);
            if replacement.is_empty() {
                migrate(&trailing, &mut out);
            } else {
                let mut words: Vec<String> =
                    replacement.iter().map(|w| (*w).to_string()).collect();
                let last = words.len() - 1;
                words[last].push_str(&trailing);
                out.extend(words);
            }
            i += match_len;
            continue;
        }
        // Rule 2: single-word drop (rule 1 guards by construction — a core
        // containing a digit or uppercase never matches the lowercase table).
        if STOPWORDS.contains(&cores[i].as_str()) {
            let trailing = trailing_punctuation(tokens[i]);
            migrate(&trailing, &mut out);
            i += 1;
            continue;
        }
        out.push(tokens[i].to_string());
        i += 1;
    }

    recapitalize_sentences(&out.join(" "))
}

/// Approximate BPE token count (§6). Integer fixed-point blend:
/// est = (3*B + 16*W + 12) / 24, B = UTF-8 bytes, W = words.
/// Mirrors Swift `TokenCompaction.estimateTokenCount` exactly.
pub fn estimate_token_count(text: &str) -> i64 {
    let word_count = text.split_whitespace().count() as i64;
    if word_count == 0 {
        return 0;
    }
    let bytes = text.len() as i64;
    (3 * bytes + 16 * word_count + 12) / 24
}

/// Fixed decorative-Unicode → ASCII map (rule 4).
pub fn normalize_unicode_punctuation(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    for ch in text.chars() {
        match ch {
            '\u{2018}' | '\u{2019}' | '\u{201B}' => out.push('\''),
            '\u{201C}' | '\u{201D}' | '\u{201F}' => out.push('"'),
            '\u{2013}' | '\u{2014}' | '\u{2015}' => out.push('-'),
            '\u{2026}' => out.push_str("..."),
            '\u{00A0}' | '\u{2007}' | '\u{202F}' => out.push(' '),
            _ => out.push(ch),
        }
    }
    out
}

/// The token's word core: leading/trailing ASCII punctuation stripped,
/// ASCII letters lowercased (ASCII-only fold). Mirrors Swift
/// `lowercaseCore(of:)`.
fn lowercase_core(token: &str) -> String {
    let chars: Vec<char> = token.chars().collect();
    let mut start = 0;
    let mut end = chars.len();
    while start < end && is_strippable_punct(chars[start]) {
        start += 1;
    }
    while end > start && is_strippable_punct(chars[end - 1]) {
        end -= 1;
    }
    chars[start..end]
        .iter()
        .map(|c| {
            if c.is_ascii_uppercase() {
                c.to_ascii_lowercase()
            } else {
                *c
            }
        })
        .collect()
}

/// The token's trailing run of clause punctuation (".,;:!?").
fn trailing_punctuation(token: &str) -> String {
    let chars: Vec<char> = token.chars().collect();
    let mut idx = chars.len();
    while idx > 0 && is_clause_punct(chars[idx - 1]) {
        idx -= 1;
    }
    chars[idx..].iter().collect()
}

/// Migrate a dropped token's trailing clause punctuation onto the
/// previous emitted token — unless it already ends with clause
/// punctuation (no doubling) or there is no previous token.
fn migrate(punctuation: &str, out: &mut [String]) {
    if punctuation.is_empty() {
        return;
    }
    let Some(last) = out.last_mut() else { return };
    let Some(last_char) = last.chars().last() else { return };
    if is_clause_punct(last_char) {
        return;
    }
    last.push_str(punctuation);
}

/// Longest phrase-table match at token position `i`.
fn phrase_match(cores: &[String], i: usize) -> Option<(usize, &'static [&'static str])> {
    let mut best: Option<(usize, &'static [&'static str])> = None;
    for (pattern, replacement) in PHRASES {
        let len = pattern.len();
        if i + len > cores.len() {
            continue;
        }
        if let Some((best_len, _)) = best {
            if len <= best_len {
                continue;
            }
        }
        let matched = (0..len).all(|k| cores[i + k] == pattern[k]);
        if matched {
            best = Some((len, *replacement));
        }
    }
    best
}

/// Sentence-start recapitalization. Mirrors Swift
/// `recapitalizeSentences(_:)` exactly: a sentence start is the string
/// start or the first non-whitespace char after ".", "!", "?" followed by
/// whitespace/end (ellipsis members and in-number dots excluded); only an
/// ASCII lowercase letter at the boundary is uppercased.
pub fn recapitalize_sentences(text: &str) -> String {
    let mut scalars: Vec<char> = text.chars().collect();
    let mut at_sentence_start = true;
    for i in 0..scalars.len() {
        let s = scalars[i];
        let is_ws = s == ' ' || s == '\t' || s == '\n' || s == '\r';
        if at_sentence_start && !is_ws {
            if s.is_ascii_lowercase() {
                scalars[i] = s.to_ascii_uppercase();
            }
            at_sentence_start = false;
        }
        if s == '.' || s == '!' || s == '?' {
            let followed_by_break = i + 1 == scalars.len()
                || matches!(scalars[i + 1], ' ' | '\t' | '\n' | '\r');
            let ellipsis_member = s == '.' && i > 0 && scalars[i - 1] == '.';
            if followed_by_break && !ellipsis_member {
                at_sentence_start = true;
            }
        }
    }
    scalars.into_iter().collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    // Mirrors Swift TokenCompactionTests case-for-case (twin-parity gate).

    #[test]
    fn drops_articles_and_fillers() {
        assert_eq!(compact("the cat sat on a mat"), "Cat sat on mat");
        assert_eq!(compact("this is really very important"), "This important");
    }

    #[test]
    fn drops_bare_copulas() {
        assert_eq!(compact("my favorite color is blue"), "My favorite color blue");
    }

    #[test]
    fn negation_survives() {
        let out = compact("I really do not want the deluxe package");
        assert!(out.contains("not"));
        assert!(!out.contains("really"));
        assert_eq!(out, "I do not want deluxe package");
    }

    #[test]
    fn contracted_negation_survives() {
        assert_eq!(
            compact("She doesn't like the very loud music"),
            "She doesn't like loud music"
        );
    }

    #[test]
    fn quantifiers_survive() {
        assert_eq!(compact("all of the tests pass"), "All of tests pass");
        assert_eq!(compact("never delete the audit log"), "Never delete audit log");
    }

    #[test]
    fn numbers_and_dates_survive() {
        assert_eq!(
            compact("The meeting moved from Tuesday, March 3rd to Thursday, March 5th"),
            "Meeting moved from Tuesday, March 3rd to Thursday, March 5th"
        );
    }

    #[test]
    fn entities_survive() {
        assert_eq!(
            compact("Sarah will send out the updated calendar invites"),
            "Sarah will send out updated calendar invites"
        );
    }

    #[test]
    fn perfect_auxiliaries_compress() {
        assert_eq!(
            compact("The meeting has been moved to Thursday"),
            "Meeting moved to Thursday"
        );
    }

    #[test]
    fn verbose_rewrites() {
        assert_eq!(compact("call me in order to confirm"), "Call me to confirm");
        assert_eq!(
            compact("delayed due to the fact that the vendor slipped"),
            "Delayed because vendor slipped"
        );
        assert_eq!(
            compact("please make sure to update your travel plans"),
            "Update your travel plans"
        );
    }

    #[test]
    fn phrase_rewrite_carries_punctuation() {
        assert_eq!(compact("The invoice has been paid."), "Invoice paid.");
    }

    #[test]
    fn unicode_punctuation_normalizes() {
        assert_eq!(
            compact("the \u{201C}quoted\u{201D} value \u{2014} right"),
            "\"quoted\" value - right"
        );
        assert_eq!(compact("wait\u{2026} done"), "Wait... done");
    }

    #[test]
    fn whitespace_collapses() {
        assert_eq!(compact("one   two\n\n\nthree\t four"), "One two three four");
        assert_eq!(compact("  padded  "), "Padded");
    }

    #[test]
    fn punctuation_migrates_on_drop() {
        assert_eq!(compact("first the, second"), "First, second");
    }

    #[test]
    fn sentence_recapitalization() {
        assert_eq!(compact("the plan works. the team agrees."), "Plan works. Team agrees.");
    }

    #[test]
    fn empty_and_determinism() {
        assert_eq!(compact(""), "");
        assert_eq!(compact("   "), "");
        assert_eq!(
            compact("The quick brown fox, really."),
            compact("The quick brown fox, really.")
        );
    }

    #[test]
    fn all_stopwords_compact_to_empty() {
        assert_eq!(compact("the a an really"), "");
    }

    // Estimator (mirrors Swift TokenCountEstimateTests)

    #[test]
    fn estimator_empty_is_zero() {
        assert_eq!(estimate_token_count(""), 0);
    }

    #[test]
    fn estimator_known_values() {
        assert_eq!(estimate_token_count("Favorite color blue."), 5);
        assert_eq!(estimate_token_count("hi"), 1);
    }

    #[test]
    fn estimator_monotone_growth() {
        let short = estimate_token_count("Quarterly planning meeting moved.");
        let long = estimate_token_count(
            "Quarterly planning meeting moved Tuesday March 3 to Thursday March 5; \
             4th floor conference room under renovation.",
        );
        assert!(long > short);
        assert!(short >= 1);
    }

    #[test]
    fn estimator_utf8_bytes() {
        assert_eq!(estimate_token_count("café"), 1);
    }
}
